### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by koaning**
@rogiervandergeer i am creating this pull request merely to get your attention and to shoot at some things. the point is that i think parts are correct, but a discussion may be helpful now. 

ive stumbled on some things that need discussion along with logging.  

1. i am proposing that we pass an `evol.Logger` object to the population apon initialisation. if you do not pass anything in then we will use the baselogger 
2. if you call `.log()` via the population/evolution then the logger that is passed in initialisation will determine what and how to log. 
3. you can see in travis what the output of a baselogger is, the current idea is to print to standardout if no file is given. it will log to a file instead if a file is given. 
3. the `evol.Logger` object is split in two verbs: `.log()` and `.handle()`. the idea is that the former will deal with the *what to handle* and the latter will deal with the *how to handle it*/*where to put it*. this should leave it free for anybody to write a custom logger but allows us to come up with a few basic ones that are general enough. we can have a `BaseLogger` to just log the population performance but offer a `PerformanceLogger` that logs the difference in time between two log moments. 
4. i am currently logging the datetime, an id for the population, an id for the individual, the fitness of every individual and the chromosome. is this well enough? 
5. there is a todo at line 202 in `population.py` -> do we still want to keep track of a populations/individuals age? i think we don't need it, it may be nicer to keep track of a generation via the logging object (ie. we keep track of how often `.log()` is called an refer to that as a generation). 
6. we can try to use pythons logging framework here, but i am wondering if we need it. opinions?

### CODE REVIEW

**Comment by koaning** [comment]
After playing around I've noticed that the current baselogger is somewhat heavy. It is *very* easy to generate a log with 23Mb if you only run a population of 200 members for 1000 iterations or so. This is not super bad (disk is cheap) but worth a mention.

File: evol/population.py [hunk 1]
```diff
@@ -214,6 +219,20 @@ def mutate(self, func, probability=1.0, **kwargs) -> 'Population':
...
 
+    def log(self, **kwargs) -> 'Population':
+        """
+        Logs a population. If a Population object was initialized with a logger
+        object then you may specify how logging is handled. The base logging 
+        operation just logs to standard out. 
+        
+        :return: self
+        """
+        logs = self.logger.log(population=self, **kwargs)
```
Comment (rogiervandergeer) [inline evol/population.py]: This feels a little redundant. Why do we need to call two methods of the logger, and doesn't it handle that by itself?

File: evol/population.py [hunk 2]
```diff
@@ -214,6 +219,20 @@ def mutate(self, func, probability=1.0, **kwargs) -> 'Population':
...
+        self.evaluate(lazy=True)
```
Comment (rogiervandergeer) [inline evol/population.py]: Why do you evaluate _after_ logging?

File: examples/number_of_parents.py [hunk 3]
```diff
@@ -39,11 +39,9 @@ def add_noise(chromosome, sigma):
            .mutate(func=add_noise, sigma=noise)
            .evaluate())
 
-    print("will start the evolutionary program, will log progress every 10 steps")
-    print(pop.maximize)
+    print("will start the evolutionary program, will log progress every step")
     for i in range(num_iter):
-        pop = pop.evolve(evo)
-        print(f"iteration:{i} best: {pop.max_individual.fitness} worst: {pop.min_individual.fitness}")
+        pop = pop.evolve(evo).log()
```
Comment (rogiervandergeer) [inline examples/number_of_parents.py]: I like this functionality!

File: evol/population.py (hunk 2 above)
Comment (koaning) [reply evol/population.py]: good call. will fix.

**Comment by koaning** [comment]
Any comments on the overal design?

File: evol/evolution.py [hunk 4]
```diff
@@ -159,6 +160,17 @@ def mutate(self, func, probability=1.0, name=None, **kwargs) -> 'Evolution':
...
 
+    def log(self, name=None) -> 'Evolution':
+        """
+        Logs a population. If a Population object was initialized with a logger
+        object then you may specify how logging is handled. The base logging 
+        operation just logs to standard out. 
+
+        :return: self
+        """
+        return self._add_step(LogStep(name))
```
Comment (rogiervandergeer) [inline evol/evolution.py]: I think we should allow kwargs here to let the user tell the logger _what_ to log. You may have multiple log steps and decide to log different things at each step.

File: evol/logger.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+import datetime as dt
+import os
+import json
+
+
+class BaseLogger():
```
Comment (rogiervandergeer) [inline evol/logger.py]: No need for the brackets

File: evol/logger.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        :return: 
+        """
+        if not self.file:
+            print(item)
+        else:
+            with open(self.file, 'a') as f:
+                f.write(item + '\n')
+
+
+class PopulationSummaryLogger(BaseLogger):
```
Comment (rogiervandergeer) [inline evol/logger.py]: I really like this functionality

File: evol/logger.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    def log(self, population, **kwargs):
+        """
+        The logger method of the Logger object determines what will be logged. 
+        :param population: 
+        :return: generator of strings to be handled
+        """
+        values = ','.join(kwargs.values())
+        for i in population:
+            string = f'{dt.datetime.now()},{population.id},{i.id},{i.fitness}' + values
+            self.handle(string)
```
Comment (rogiervandergeer) [inline evol/logger.py]: I think it is much better that we handle the `handle` internally, such that the `Population` only makes a single call. I think we may take it even one step further: have `log()` call two methods:

```
def log(self, population, **kwargs):
    lines = self.format(population, **kwargs)
    self.handle(lines)
```

This way a user can easily inherit from our loggers and overwrite `format`. He/she doesn't need to think about calling `handle`; only to return what needs to be logged.

File: evol/population.py [hunk 8]
```diff
@@ -217,6 +222,18 @@ def mutate(self, func, probability=1.0, **kwargs) -> 'Population':
             individual.mutate(func, probability=probability, **kwargs)
         return self
 
+    def log(self, **kwargs) -> 'Population':
```
Comment (rogiervandergeer) [inline evol/population.py]: Should we allow the user to _not_ evaluate before logging? We could have a flag that allows that

File: examples/simple_logging.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+pop = Population(chromosomes=[random_start() for _ in range(200)],
+                 eval_function=func_to_optimise,
+                 maximize=True,
+                 logger=BaseLogger(file="/tmp/evol.log"))
+
+evo1 = (Evolution()
+       .survive(fraction=0.1)
+       .breed(parent_picker=pick_random_parents, combiner=make_child)
+       .mutate(func=add_noise, sigma=0.2)
+       .log())
```
Comment (rogiervandergeer) [inline examples/simple_logging.py]: We should differentiate what we log here between the two evolutions, as an example

**Comment by rogiervandergeer** [comment]
Concerning the points in the opening post:

1. Agreed
2. I think the `.log()` call should accept arguments on what to log. I'm thinking along the lines of `log(summary=True, all_individuals=False)`. 
3. Great
4. I like it, with the changes we talked about before
5. That might be too much. I'm not sure what is reasonable to log. Perhaps logging chromosomes is too much? That feels like a job for the serializer. Logging _the best_ chromosome might be a good feature though.
6. Keeping track of a population's age through the logger makes sense. But you may want to know how long a certain individual has been around. I'm not sure.
7. I think we should give it a try. It feels like a natural thing to do. On the other hand 'simply' logging to file or stdout should certainly be an option.

File: evol/population.py (hunk 8 above)
Comment (koaning) [reply evol/population.py]: This may add confusion. What would be the base behaviour? If an individual is not evaluated we do not log it?

**Comment by koaning** [comment]
Python's `logging` framework allows for simple file as well as stdout. It also has some support for threading if multiple processes are writing to a file (which is gonna happen if we ever explore parallel). So I'll focus on that for the next iteration. 

Implicitly you can track the age by logging every iteration and checking if the individuals id is still there. I'll ignore the age for now since you can always derive it. 

I'll remove the chromosome from the `baselogger`, it feels like the checkpointing mechanism needs to worry about this more than the logging mechanism.

**Comment by koaning** [comment]
Mhm. Something is up. Ever since I've merged some changes in `master` into the `logging` branch the build seems to be failing. Not even the base examples will run: 

https://travis-ci.org/godatadriven/evol/builds/323565816?utm_source=github_status&utm_medium=notification

When investigating it seems that pythons logging module does not play nice with a `deepcopy` (see [here](https://github.com/ipython/ipython/issues/8023)). I've changed line 83 in `population.py` to confirm that this removes the error. Pythons logging framework makes use of thread programming to make the logging framework robust against multiple writes to a single file, this does not seem to play nice with a `deepcopy`. I've changed the `deepcopy` into a normal `copy` but we should discuss this.

**Comment by rogiervandergeer** [comment]
That is a problem; a normal copy is not sufficient. There is a solution in the islands branch, which implements a proper __copy__ of the population. 

My guess is that the deepcopy also copies the logger.

**Comment by koaning** [comment]
So I've copied the island copy, which I like very much, in the hope that this is enough as a first version to get logging to work. @rogiervandergeer let me know what you think.

File: evol/population.py [hunk 10]
```diff
@@ -27,16 +27,23 @@ class Population:
...
+        self.generation = generation
         self.individuals = [Individual(chromosome=chromosome) for chromosome in chromosomes]
-        self.intended_size = len(chromosomes)
+        self.intended_size = len(chromosomes) if intended_size is None else intended_size
         self.maximize = maximize
         self.logger = logger
 
+    def __copy__(self):
+        result = self.__class__(chromosomes=self.chromosomes,
+                               eval_function=self.eval_function,
```
Comment (rogiervandergeer) [inline evol/population.py]: Indentation seems to go wrong here

**Pull request merged**